src/blockdata/witness.rs

```diff
@@ -0,0 +1,328 @@
+        Witness {
+            content: Vec::new(),
+            witness_elements: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for WitnessIterator<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let vec = &self.witness.content;
+        if self.cursor >= vec.len() {
+            None
+        } else {
+            let var = VarInt::consensus_decode(&vec[self.cursor..])
+                .expect("is granted witness.content contains varint because created only from internal methods");
+            let start = self.cursor + var.len();
+            let end = start + var.0 as usize;
+            self.cursor = end;
+            Some(&vec[start..end])
+        }
+    }
+}
+
+// Serde keep backward compatibility with old Vec<Vec<u8>> format
+#[cfg(feature = "serde")]
+impl serde::Serialize for Witness {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let vec: Vec<_> = self.iter().map(|e| e.to_vec()).collect();
+        serde::Serialize::serialize(&vec, serializer)
+    }
+}
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Witness {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let vec: Vec<Vec<u8>> = serde::Deserialize::deserialize(deserializer)?;
+        Ok(vec.into())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use blockdata::witness::Witness;
+    use consensus::{deserialize, serialize};
+    use hashes::hex::{FromHex, ToHex};
+    use Transaction;
+
+    #[test]
+    fn test_push() {
+        let mut witness = Witness::default();
+        witness.push(&vec![0u8]);
+        let expected = Witness {
+            witness_elements: 1,
+            content: vec![1u8, 0],
+        };
+        assert_eq!(witness, expected);
+        witness.push(&vec![2u8, 3u8]);
+        let expected = Witness {
+            witness_elements: 2,
+            content: vec![1u8, 0, 2, 2, 3],
+        };
+        assert_eq!(witness, expected);
+    }
+
+    #[test]
+    fn test_witness() {
+        let w0 =
+            Vec::from_hex("03d2e15674941bad4a996372cb87e1856d3652606d98562fe39c5e9e7e413f2105")
+                .unwrap();
+        let w1 = Vec::from_hex("000000").unwrap();
+        let witness_vec = vec![w0, w1];
+        let witness_serialized: Vec<u8> = serialize(&witness_vec);
+        let witness = Witness {
+            content: witness_serialized[1..].to_vec(),
+            witness_elements: 2,
+        };
+        for (i, el) in witness.iter().enumerate() {
+            assert_eq!(witness_vec[i], el);
+        }
+
+        let w_into: Witness = witness_vec.into();
+        assert_eq!(w_into, witness);
+
+        assert_eq!(witness_serialized, serialize(&witness));
+
+        //assert_eq!(32, std::mem::size_of::<Witness>());
+        //assert_eq!(24, std::mem::size_of::<Option<Vec<u8>>>());
```

Comment:
nit: remove comments here?

Reply:
They were removed in latest commit, but present in old one, after the squash they are removed completely